internal/core/services/autoscaler/autoscaler.go

```diff
@@ -0,0 +1,88 @@
+// MIT License
+//
+// Copyright (c) 2021 TFG Co
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in all
+// copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+package autoscaler
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/topfreegames/maestro/internal/core/entities"
+	"github.com/topfreegames/maestro/internal/core/entities/autoscaling"
+	"github.com/topfreegames/maestro/internal/core/ports"
+	autoscalerPorts "github.com/topfreegames/maestro/internal/core/ports/autoscaler"
+	"github.com/topfreegames/maestro/internal/core/services/autoscaler/policies/roomoccupancy"
+)
+
+// PolicyFactory defines a factory to be used by the autoscaler.
+// By default it uses the policyFactory inside the same struct.
+type PolicyFactory func(autoscaling.PolicyType) (autoscalerPorts.Policy, error)
+
+// Autoscaler is a service that holds dependencies to execute autoscaling feature.
+type Autoscaler struct {
+	roomStorage ports.RoomStorage
+	PolicyFactory
+}
+
+// NewAutoscaler returns a new instance of autoscaler.
+func NewAutoscaler(roomStorage ports.RoomStorage) *Autoscaler {
+	autoscaler := &Autoscaler{
+		roomStorage: roomStorage,
+	}
+
+	autoscaler.PolicyFactory = autoscaler.DefaultPolicyFactory
+
+	return autoscaler
+}
+
+// CalculateDesiredNumberOfRooms return in success case the number of rooms that a Scheduler should have based on its policy,
+// in error cases it returns nil and the error.
```

Comment:
```suggestion
// CalculateDesiredNumberOfRooms return the number of rooms that a Scheduler should have based on its policy or error if it can calculate.
```